### PR TITLE
#169258648 User can search trip requests by preference

### PR DESCRIPTION
**Pull request by nziokaivy**
#### What does this PR do?

As a user, I should be able to use the search component so, I can easily retrieve records from both the request and approval table

#### Description of Task to be completed?

- [x] Add tests
- [x] Add verify token and is verified validation
- [x] Add validate query parameters validation middleware
- [x] Add search query parameter query function
- [x] Add search controller
- [x] Add route API documentation

#### How should this be manually tested?
- Clone this repository using the command: `git clone https://github.com/andela/team-odd-bn-backend.git`
- Checkout to this branch using the command: `git checkout ft-search-functionality-169258648`
- Run the application using the command: npm run dev-start
- Run the tests using the command: npm test

#### Any background context you want to provide?

As requests and approval tables grow, it should be easy to find requests through the search functionality

#### What are the relevant pivotal tracker stories?
[#169258648](https://www.pivotaltracker.com/story/show/169258648)

###### Search by originId
image

###### Search by destinationId
image

###### Search by trip type
image

###### Search by status
image

###### Search by trip owner(by their first name)
image

#### Questions:

### CODE REVIEW

File: src/services/commonSearchQueries.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
+import { Op } from 'sequelize';
+import {
+  tripRequests, status, tripTypes, users
+} from '../database/models';
+
+
+const query = (searchQueryParams) => ({
+  where: {
+    ...searchQueryParams.startDate && {
+      startDate: {
```
Comment (victkarangwa) [inline src/services/commonSearchQueries.js]: @ivymwende I think we need to have a limitation in fetching data from DB. I found that any authenticated user can get all the data he/she may want that are not related to him/her.

**Review by victkarangwa — CHANGES_REQUESTED** [review]

**Review by victkarangwa — APPROVED** [review]

File: src/middlewares/validateSearchQueries.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    }
+    return false;
+  });
+
+  const invalidKeys = searchKeys.filter(key => key === false);
+  if (invalidKeys.length > 0) {
+    return Response.errorMessage(req, res, 'You provided an invalid search query key(s) parameter.Your search key should be either originId, destinationId, startDate, returnDate, firstName, status, or tripType', 400);
+  }
+
+  if (query.originId === '' || query.destinationId === '' || query.startDate === '' || query.returnDate === '' || query.firstName === '' || query.tripType === '') {
```
Comment (hezronkimutai) [inline src/middlewares/validateSearchQueries.js]: Hi @ivymwende , lets make this line shorter ie 
```
 if (query.originId === '' || query.destinationId === '' 
      || query.startDate === '' || query.returnDate === '' 
       || query.firstName === '' || query.tripType === ''
)
```

File: src/services/commonSearchQueries.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+          },
+          attributes: ['firstName'],
+        }
+      ]
+    }
+  ],
+});
+
+
+export default query;
```
Comment (hezronkimutai) [inline src/services/commonSearchQueries.js:76]: Wow! this  file looks nice.

**Review by hezronkimutai — CHANGES_REQUESTED** [review]
Nice job @ivymwende , I left a comment, please view it and we will be good to go.

**Review by hezronkimutai — APPROVED** [review]

File: src/services/SearchService.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ * @description GET /api/search
+ * @param {object} req request object
+ * @memberof SearchService
+ * @returns {object} data
+ */
+  static async searchByPreference(req) {
+    const { query } = req;
+    const searchingTrips = await CommonQueries.findAll(
+      trips, commonSearchQueries(query, req.user.id)
+    );
```
Comment (william000000) [inline src/services/SearchService.js:21]: Wonderful nested queries! That's great.

File: src/services/commonSearchQueries.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+      originId: {
+        [Op.eq]: parseInt((searchQueryParams.originId), 10)
+      }
+    },
+    ...searchQueryParams.destinationId && {
+      destinationId: {
+        [Op.eq]: parseInt((searchQueryParams.destinationId), 10)
+      }
+    },
+  },
```
Comment (william000000) [inline src/services/commonSearchQueries.js:29]: I like the way you come up with this queries.

**Review by william000000 — APPROVED** [review]

**Review by william000000 — APPROVED** [review]
Great Job @nziokaivy. I have lost a mistake that you might have done.

**Review by julietezekwe — APPROVED** [review]
